File: app/src/androidTest/java/com/ukdev/smartbuzz/robots/SetupActivityRobot.java
package com.ukdev.smartbuzz.robots;

import android.app.Activity;
import android.content.Intent;
import android.support.test.espresso.assertion.ViewAssertions;
import android.support.test.espresso.matcher.ViewMatchers;
import android.support.test.rule.ActivityTestRule;

import com.ukdev.smartbuzz.R;
import com.ukdev.smartbuzz.misc.IntentExtra;

import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.action.ViewActions.click;
import static android.support.test.espresso.matcher.ViewMatchers.withId;

public class SetupActivityRobot {

    private ActivityTestRule<? extends Activity> rule;

    public SetupActivityRobot(ActivityTestRule<? extends Activity> rule) {
        this.rule = rule;
    }

    public SetupActivityRobot launchActivity() {
        return launchActivity(false);
    }

    public SetupActivityRobot launchActivityInEditMode() {
        return launchActivity(true);
    }

    public SetupActivityRobot assertDeleteActionButtonIsVisible() {
        onView(withId(R.id.item_delete)).check(ViewAssertions.matches(ViewMatchers.isDisplayed()));
        return this;
    }

    public SetupActivityRobot clickOnSave() {
        onView(withId(R.id.fab_setup)).perform(click());
        return this;
    }

    public SetupActivityRobot clickOnDelete() {
        onView(withId(R.id.item_delete)).perform(click());
        return this;
    }

    private SetupActivityRobot launchActivity(boolean editMode) {
        rule.launchActivity(new Intent().putExtra(IntentExtra.EDIT_MODE.toString(), editMode));
        return this;
    }

}

File: app/src/androidTest/java/com/ukdev/smartbuzz/robots/SetupActivityRobot.kt
package com.ukdev.smartbuzz.robots

import android.support.test.InstrumentationRegistry
import android.support.test.espresso.Espresso.onView
import android.support.test.espresso.action.ViewActions.click
import android.support.test.espresso.intent.Intents.intended
import android.support.test.espresso.intent.matcher.IntentMatchers.hasComponent
import android.support.test.espresso.matcher.ViewMatchers.withId
import android.support.test.rule.ActivityTestRule
import com.ukdev.smartbuzz.R
import com.ukdev.smartbuzz.activities.MainActivity
import com.ukdev.smartbuzz.activities.SetupActivity
import org.junit.Rule

class SetupActivityRobot : BaseActivityRobot() {

    @Rule
    @JvmField
    val rule = ActivityTestRule<SetupActivity>(SetupActivity::class.java,
                                              initialTouchMode, launchActivity)

    fun launchActivity(): SetupActivityRobot {
        val context = InstrumentationRegistry.getTargetContext()
        val intent = SetupActivity.getIntent(context)
        rule.launchActivity(intent)
        return this
    }

    fun launchActivity(alarmId: Int): SetupActivityRobot {
        val context = InstrumentationRegistry.getTargetContext()
        val intent = SetupActivity.getIntent(context, alarmId)
        rule.launchActivity(intent)
        return this
    }

    fun clickOnSave(): SetupActivityRobot {
        onView(withId(R.id.fab_setup)).perform(click())
        return this
    }

    fun checkIfLeadsToMainActivity(): SetupActivityRobot {
        intended(hasComponent(MainActivity::class.java.name))
        return this
    }

}